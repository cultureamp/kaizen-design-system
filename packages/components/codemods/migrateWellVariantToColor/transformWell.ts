import fs from "fs"
import path from "path"
// eslint-disable-next-line import/no-extraneous-dependencies
import ts from "typescript"
import { transformSource } from "../utils"

/** Recurses through AST to find the import specifier name and check it against the `importSpecifierTarget`. If found, it will return the import name or alias, otherwise will return `undefined` */
export const getImportAlias = (
  node: ts.Node,
  importSpecifierTarget: string
): string | undefined => {
  let alias: string | undefined

  const visitNode = (visitedNode: ts.Node): string | undefined => {
    if (ts.isImportDeclaration(visitedNode)) {
      const moduleSpecifier = visitedNode.moduleSpecifier.getText()
      if (moduleSpecifier.includes("@kaizen/components")) {
        const namedBindings = visitedNode.importClause?.namedBindings
        if (namedBindings && ts.isNamedImports(namedBindings)) {
          namedBindings.elements.forEach(importSpecifier => {
            const importName = importSpecifier.name.getText()
            const importAlias = importSpecifier.propertyName
              ? importSpecifier.propertyName.getText()
              : importName
            if (importAlias === importSpecifierTarget) {
              alias = importName
            }
          })
        }
      }
    }
    return alias || ts.forEachChild(visitedNode, visitNode) || undefined
  }

  return visitNode(node)
}

/** a  helper function to get the initializer text from   */
const getInitializerText = (
  initializer: ts.JsxAttributeValue
): string | undefined => {
  if (ts.isStringLiteral(initializer)) {
    return initializer.text
  }

  const expression = ts.isJsxExpression(initializer) && initializer.expression

  if (expression) {
    const expressionText = expression.getText()

    if (expressionText.match(/^['"`]/)) {
      return expressionText.replace(/^['"`]|['"`]$/g, "")
    }
  }
  return undefined
}

/** Recurses through AST to find and update any jsx element that matched the importAlias */
export const wellTransformer =
  (context: ts.TransformationContext, importAlias: string) =>
  (rootNode: ts.Node): ts.Node => {
    function visit(node: ts.Node): ts.Node {
      if (ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node)) {
        if (node.tagName.getText() === importAlias) {
          let hasVariant = false
          let hasColor = false
          let newAttributes = node.attributes.properties.map(attr => {
            if (ts.isJsxAttribute(attr) && attr.name.getText() === "variant") {
              hasVariant = true
              const valueName =
                attr.initializer && getInitializerText(attr.initializer)

              if (valueName) {
                let colorValue: string = "gray"
                switch (valueName) {
                  case "default":
                    colorValue = "gray"
                    break
                  case "informative":
                    colorValue = "blue"
                    break
                  case "cautionary":
                    colorValue = "yellow"
                    break
                  case "assertive":
                    colorValue = "orange"
                    break
                  case "negative":
                    colorValue = "red"
                    break
                  case "positive":
                    colorValue = "green"
                    break
                  case "prominent":
                    colorValue = "purple"
                    break
                  default:
                    colorValue = "gray"
                }
                return ts.factory.createJsxAttribute(
                  ts.factory.createIdentifier("color"),
                  ts.factory.createStringLiteral(colorValue)
                )
              }
            }
            if (ts.isJsxAttribute(attr) && attr.name.getText() === "color") {
              hasColor = true
            }
            return attr
          })

          if (!hasColor && !hasVariant) {
            newAttributes = [
              ...newAttributes,
              ts.factory.createJsxAttribute(
                ts.factory.createIdentifier("color"),
                ts.factory.createStringLiteral("gray")
              ),
            ]
          }

          if (ts.isJsxOpeningElement(node)) {
            return ts.factory.updateJsxOpeningElement(
              node,
              node.tagName,
              node.typeArguments,
              ts.factory.createJsxAttributes(newAttributes)
            )
          } else if (ts.isJsxSelfClosingElement(node)) {
            return ts.factory.updateJsxSelfClosingElement(
              node,
              node.tagName,
              node.typeArguments,
              ts.factory.createJsxAttributes(newAttributes)
            )
          }
        }
      }
      return ts.visitEachChild(node, visit, context)
    }
    return ts.visitNode(rootNode, visit)
  }

/** runs the transformer and writes the updated source back to the path provided */
export const updateFileContents = (
  sourceFile: ts.SourceFile,
  importAlias: string
): string => {
  const updatedSourceFile = transformSource(
    sourceFile,
    wellTransformer,
    importAlias
  )

  const printer = ts.createPrinter()
  const updatedSource = printer.printFile(updatedSourceFile)

  return updatedSource
}

/** Walks the directory given and runs the runs the AST updater */
export const processDirectory = (dir: string): void => {
  if (dir.includes("node_modules")) {
    return
  }

  const files = fs.readdirSync(dir)
  files.forEach(file => {
    const fullPath = path.join(dir, file)
    if (fs.statSync(fullPath).isDirectory()) {
      processDirectory(fullPath)
    } else if (fullPath.endsWith(".tsx")) {
      const source = fs.readFileSync(fullPath, "utf8")
      const sourceFile = ts.createSourceFile(
        fullPath,
        source,
        ts.ScriptTarget.Latest,
        true
      )
      const importAlias = getImportAlias(sourceFile, "Well")

      if (importAlias) {
        const updatedSourceFile = updateFileContents(sourceFile, importAlias)

        fs.writeFileSync(fullPath, updatedSourceFile, "utf8")
      }
    }
  })
}
