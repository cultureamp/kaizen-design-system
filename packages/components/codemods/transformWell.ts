import fs from "fs"
import path from "path"
import ts from "typescript"

/** Recurses through AST to determine if a valid import statement for Well has been provided */
const checkForValidImports = (node: ts.Node): boolean => {
  let shouldTransform = false

  const visitNode = (visitedNode: ts.Node): boolean => {
    if (ts.isImportDeclaration(visitedNode)) {
      const moduleSpecifier = visitedNode.moduleSpecifier.getText()
      if (moduleSpecifier.includes("@kaizen/components")) {
        const namedBindings = visitedNode.importClause?.namedBindings
        if (namedBindings && ts.isNamedImports(namedBindings)) {
          namedBindings.elements.forEach(importSpecifier => {
            if (importSpecifier.name.getText() === "Well") {
              shouldTransform = true
            }
          })
        }
      }
    }
    return shouldTransform || ts.forEachChild(visitedNode, visitNode) || false
  }

  return visitNode(node)
}

/** Transformer function to update the inputs of Well to the new colors */
const wellTransformer =
  <T extends ts.Node>(context: ts.TransformationContext) =>
  (rootNode: T) => {
    function visit(node: ts.Node): ts.Node {
      if (ts.isJsxOpeningElement(node)) {
        // TODO: check alias names for Well
        if (node.tagName.getText() === "Well") {
          let hasVariant = false
          let newAttributes = node.attributes.properties.map(attr => {
            if (ts.isJsxAttribute(attr) && attr.name.getText() === "variant") {
              hasVariant = true
              if (attr.initializer && ts.isStringLiteral(attr.initializer)) {
                let colorValue: string = "gray"
                switch (attr.initializer.text) {
                  case "default":
                    colorValue = "gray"
                    break
                  case "informative":
                    colorValue = "blue"
                    break
                  case "cautionary":
                    colorValue = "yellow"
                    break
                  case "assertive":
                    colorValue = "orange"
                    break
                  case "negative":
                    colorValue = "red"
                    break
                  case "positive":
                    colorValue = "green"
                    break
                  case "prominent":
                    colorValue = "purple"
                    break
                  default:
                    colorValue = "gray"
                }
                return ts.factory.createJsxAttribute(
                  ts.factory.createIdentifier("color"),
                  ts.factory.createStringLiteral(colorValue)
                )
              }
            }
            return attr
          })

          if (!hasVariant) {
            newAttributes = [
              ...newAttributes,
              ts.factory.createJsxAttribute(
                ts.factory.createIdentifier("color"),
                ts.factory.createStringLiteral("gray")
              ),
            ]
          }

          return ts.factory.updateJsxOpeningElement(
            node,
            node.tagName,
            node.typeArguments,
            ts.factory.createJsxAttributes(newAttributes)
          )
        }
      }
      return ts.visitEachChild(node, visit, context)
    }
    return ts.visitNode(rootNode, visit)
  }

/** Returns original the source file if invalid, otherwise runs the well transformer  */
export const transformWellSource = (
  sourceFile: ts.SourceFile
): ts.SourceFile => {
  const shouldTransform = checkForValidImports(sourceFile)

  if (!shouldTransform) {
    return sourceFile
  }
  const result = ts.transform(sourceFile, [wellTransformer])
  const transformedSource = result.transformed[0] as ts.SourceFile

  return transformedSource
}

/** runs the transformer and writes the updated source back to the path provided */
export const updateFileContents = (filePath: string): string => {
  const source = fs.readFileSync(filePath, "utf8")
  const sourceFile = ts.createSourceFile(
    filePath,
    source,
    ts.ScriptTarget.Latest,
    true
  )
  const updatedSourceFile = transformWellSource(sourceFile)

  const printer = ts.createPrinter()
  const updatedSource = printer.printFile(updatedSourceFile)

  return updatedSource
}

/** Walks the directory given and runs the runs the AST updater */
export const processDirectory = (dir: string): void => {
  const files = fs.readdirSync(dir)
  files.forEach(file => {
    const fullPath = path.join(dir, file)
    if (fs.statSync(fullPath).isDirectory()) {
      processDirectory(fullPath)
    } else if (fullPath.endsWith(".tsx")) {
      fs.writeFileSync(fullPath, updateFileContents(fullPath), "utf8")
    }
  })
}
