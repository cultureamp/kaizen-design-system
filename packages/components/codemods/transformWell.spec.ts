import path from "path"
import ts from "typescript"
import { transformWellSource, updateFileContents } from "./transformWell"

// This function is used to parse a stringified JSX element into an AST
function parseJsx(jsx: string): ts.SourceFile {
  return ts.createSourceFile(
    "tempFile.tsx",
    jsx,
    ts.ScriptTarget.Latest,
    true,
    ts.ScriptKind.TSX
  )
}

// This function is used to print the AST as a string and is only used for testing
function printAst(ast: ts.SourceFile): string {
  const printer = ts.createPrinter()
  return printer.printFile(ast)
}

// describe("transformWellSource", () => {
//   it("should only transform when Well is imported from @kaizen/components", () => {
//     const input = parseJsx(`
//       import { Well } from "@kaizen/draft-well"
//       export const TestComponent = () => <div><Well>Test</Well></div>`)
//     const output = parseJsx(`
//       import { Well } from "@kaizen/draft-well"
//       export const TestComponent = () => <div><Well>Test</Well></div>`)
//     const transformed = transformWellSource(input)
//     expect(printAst(transformed).trim()).toBe(printAst(output).trim())
//   })

//   it('should replace variant="default" with color="gray"', () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well variant="default">Test</Well>
//       `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well color="gray">Test</Well>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it('should replace variant="informative" with color="blue"', () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well variant="informative">Test</Well>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well color="blue">Test</Well>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it('should add color="gray" if variant is not specified', () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well>Test</Well>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well color="gray">Test</Well>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it("should handle multiple attributes and replace only variant", () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well variant="informative" id="123">Test</Well>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well color="blue" id="123">Test</Well>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it("should handle multiple attributes and add color if variant is not specified", () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well id="123">Test</Well>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <Well id="123"  color="gray">Test</Well>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it("should handle nested Well components", () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <div><Well id="123">Test</Well></div>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <div><Well id="123" color="gray">Test</Well></div>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })

//   it("should transform multiple Wells", () => {
//     const inputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <div><Well variant="informative">Test</Well><Well>Test 2</Well></div>
//     `)
//     const outputAst = parseJsx(`
//       import {Well} from "@kaizen/components"
//       export const TestComponent = () => <div><Well color="blue">Test</Well><Well color="gray">Test 2</Well></div>
//     `)
//     const transformed = transformWellSource(inputAst)
//     expect(printAst(transformed).trim()).toBe(printAst(outputAst).trim())
//   })
// })

describe("updateFileContents", () => {
  it("should update Well components imported from @kaizen/components", async () => {
    const filePath = path.resolve(path.join(__dirname, "./fixtures/Well.tsx"))

    const updatedFileContent = updateFileContents(filePath)

    expect(updatedFileContent).toMatchSnapshot()
  })
  it("should not update Well components imported from @kaizen/draft-well or legacy packages", async () => {
    const legacyFilePath = path.resolve(
      path.join(__dirname, "./fixtures/LegacyWell.tsx")
    )

    const updatedFileContent = updateFileContents(legacyFilePath)
    expect(updatedFileContent).toMatchSnapshot()
  })
  it("should update Well components imported from @kaizen/component/path", async () => {
    const legacyFilePath = path.resolve(
      path.join(__dirname, "./fixtures/WellV3.tsx")
    )

    const updatedFileContent = updateFileContents(legacyFilePath)
    expect(updatedFileContent).toMatchSnapshot()
  })
})
