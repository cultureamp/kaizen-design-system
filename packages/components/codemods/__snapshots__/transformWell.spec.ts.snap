// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateFileContents should not update Well components imported from @kaizen/draft-well or legacy packages 1`] = `
"import * as React from "react";
import { Well } from "kaizen/draft-well";
export const TestWellComponentUnhappy = (): JSX.Element => (<div>
    <Well variant="default">Test 1 should have default variant</Well>
    <Well>Test 2 should have no change</Well>
    <Well>Test 3 should have informative variant</Well>
    <Well id="test-id" variant="informative">
      Test 3 should have informative variant and an id prop
    </Well>
  </div>);
"
`;

exports[`updateFileContents should update Well components imported from @kaizen/component/path 1`] = `
"import * as React from "react";
import { Well } from "@kaizen/components/v3";
export const TestWellComponent = (): JSX.Element => (<div>
    <Well color="gray">Test 1 should have color gray</Well>
    <Well color="gray">Test 2 should have color gray</Well>
    <Well color="blue">Test 3 should have color blue</Well>
    <Well id="test-id" color="blue">
      Test 3 should have color blue and an id prop
    </Well>
  </div>);
"
`;

exports[`updateFileContents should update Well components imported from @kaizen/components 1`] = `
"import * as React from "react";
import { Well } from "@kaizen/components";
export const TestWellComponent = (): JSX.Element => (<div>
    <Well color="gray">Test 1 should have color gray</Well>
    <Well color="gray">Test 2 should have color gray</Well>
    <Well color="blue">Test 3 should have color blue</Well>
    <Well id="test-id" color="blue">
      Test 3 should have color blue and an id prop
    </Well>
  </div>);
"
`;
